### DISCUSSION TITLE
Change sheets vs. Migrations

### DISCUSSION THREAD

**Discussion by brynnz22** [microbiomedata/issues#807] [General]
In the migration squad we have discussed only doing migrations when schema changes create validation issues in the existing data in Mongo. 

Change sheets are tricky to create and use, especially when changing large amounts of data or nested data. Some change sheets, though only changing a few documents, can take much longer to get right and pass validation than writing a migrator would. 

@kheal @SamuelPurvine and I have been working on updating the new `MassSpectrometry` instances (previously the generic `OmicsProcessing` records for lipids, metabolomics, proteomics, and nom) to add additional information that was created in Berkeley (e.g. `eluent_intorduction`, `has_mass_spec_config`, `has_chromatographic_config`, and `has_calibration`). Though the schema changes in Berkeley did not invalidate any data because all the changes are currently optional - this change sheet is roughly 8,500 rows - I have not been able to validate using Berkeley runtime yet because it is too large. It will have to be broken up into chunks; as @bmeluch has mentioned she breaks them up into chunks of 200 rows, which will be about 43 submissions for the MassSpectrometry change sheet. 

Is it worth considering using migrators to change data in Mongo to ease this process? Or are there other ideas or thoughts? This has been discussed with @eecavanna and @turbomam in the Migration squad, but we wanted to pull it out into a larger discussion given the complexity of the Berkeley roll out. 

Please tag anyone else!

@mslarae13 @lamccue @aclum @pkalita-lbl

**Comment by eecavanna** [comment]
**Regarding changesheet size limits:** Assuming updating the changesheet API to tolerate larger changesheets is not on the table, I think I'd prefer writing a wrapper script to handle breaking up large changesheets into smaller ones and submitting the smaller ones; over using a migrator for data insertion that isn't related to schema changes.

The script would take as its input:

- Runtime API credentials
- Original changesheet file

The script would make as its output:

- Partial changesheet files created by dividing the original one (for the user's reference/debug)
- HTTP responses corresponding to the validation of the partial changesheet files
- HTTP responses corresponding to the submission of the partial changesheet files

The script would:

1. Divide the original changesheet into partial ones
2. Authenticate an HTTP client with the Runtime API
3. Use the Runtime API to validate each changesheet
4. Use the Runtime API to submit each changesheet

**Reply by bmeluch** [reply]
possibly relevant, example of my bash commands for dealing with large changesheets - I run it in pieces to check dev results before submitting to prod
https://github.com/microbiomedata/issues/issues/612#issuecomment-2239725442

**Comment by mslarae13** [comment]
As a note, change-sheet was **never** supposed to be a long term solution. We were always supposed to replace this process.

Curious on why (having not yet read @eecavanna 's comment), we wouldn't want to just transition to using migrations.

**Reply by bmeluch** [reply]
What's the learning curve like on migrations vs. changesheets?

**Reply by brynnz22** [reply]
I think migrators might be a slightly higher learning curve, but once you have written a few they are much quicker than the change sheets for nested changes or large changes.

**Comment by dwinston** [comment]
If folks are comfortable preparing JSON for API requests, `POST /metadata/json:submit` can be used for metadata updates, not just inserts of "new" documents. However, the entirety of an updated document needs to be supplied, as the db will do a “full replace”.

The motivation for changesheets IIRC was "it's hard to format JSON, I'm more comfortable with formatting a spreadsheet".

**Comment by turbomam** [comment]
Can somebody please share the changesheet here? or the first 20 rows?

**Reply by brynnz22** [reply]
Hey Mark here is a folder from the study refresh squad with a bunch of different kinds of change sheets: https://drive.google.com/drive/folders/13xormRHMnY3Ui6dwwPJf6-hbqqjoCwLy

**Reply by brynnz22** [reply]
[change_sheet.csv](https://github.com/user-attachments/files/16461980/change_sheet.csv)

Here is a link to the full change sheet. I guess it is 13,200 rows!

**Comment by mslarae13** [comment]
@brynnz22 & @bmeluch & @kheal & @aclum  Or anyone else

Can you comment the pain points with change sheet?
Discussion around what we do in the future... new tool or make change sheet better and knowing how it's going and what the issues would be helpful!

**Reply by bmeluch** [reply]
- when dealing with multiple change sheets on the command line (ie when breaking up long ones) I have had occasions where only some of the sheets seemed to actually go through and I have to resubmit - this is mostly fixed by adding `sleep` to avoid submitting commands to the API too quickly, but does still sometimes happen
- the eternal DOS/UNIX whitespace problems if switching between editors (this may just be an issue for me and my environment though)
- formatting multivalued slots

I do like looking at a spreadsheet more than JSON though.

**Reply by brynnz22** [reply]
- I would add a big pain point is updating/removing/inserting into nested slots. It gets a complex to update.
- I wouldn't mind using json in the future, but it seems scary to have to replace the full document.

**Comment by dwinston** [comment]
FYI `nginx.ingress.kubernetes.io/proxy-body-size` is currently set to `25m` (so 25 MB) for the nmdc.api ingress on rancher, so that's a practical limit. We can bump this up.

**Comment by PeopleMakeCulture** [comment]
The changesheets hackathon team is creating a Google doc here to understand: https://docs.google.com/document/d/1rxRQuJI8W-HCoJGiFFnXVIRfyRvod7NUpr2Myo7NtsM/edit?usp=sharing

**Comment by aclum** [comment]
I have a strong preference for using either changesheet:submit or queries:run to make changes to data over ad hoc migrations as it is easier to track when data was changed or deleted since that creates a logging trail. For those that prefer json format queries:run would my recommendation. queries:run would be easier than json:submit b/c you wouldn't have to have the entire document, just the part you want to change.

**Reply by brynnz22** [reply]
Ah great to know. I know you've mentioned queries:run before. I will try it out.

**Comment by eecavanna** [comment]
During Friday's hackathon, @PeopleMakeCulture and I brainstormed a hypothetical standalone tool team members could use to transform changesheets into `/queries:run` request payloads. I started building a proof of concept over the weekend. Here's what it looks like as of now: https://eecavanna.github.io/nmdc-changesheet-converter/

**Reply by eecavanna** [reply]
I updated this PoC so that it can also be used to _split_ one large changesheet into multiple smaller ones.

**Comment by eecavanna** [comment]
We talked about the state of this Discussion during today's migration meeting. I volunteered to document some task-specific changesheets snippets, as an incremental step toward making changesheets easier for people to use. I'll track that task in the following ticket: https://github.com/microbiomedata/nmdc-runtime/issues/832
